File: frontend/src/types/train.ts
/**
 * TypeScript mirrors of the Phase 189/190 Train API Pydantic schemas
 * (app/schemas/train.py). Field-for-field, literal unions instead of bare
 * `string` per CLAUDE.md's type-safety rule.
 *
 * `TrainPuzzle` carries no answer key (POOL-10 / P-01, LOCKED).
 * `last_move_uci` (190-02, SOLV-02) is the position's arrival — the prior
 * half-move — never what to play next, so it does not reopen POOL-10. Do
 * not add fields here without re-reading that decision.
 */

import type { TrainMoveTier } from '@/lib/trainScore';

export interface TrainPuzzle {
  position: number;
  game_id: number | null;
  ply: number;
  fen: string;
  side_to_move: 'white' | 'black';
  last_move_uci: string | null;
}

/**
 * One recorded solve's outcome, part of `TrainSessionResponse.solved_results`
 * (quick task 260728-tgc, BUGFIX-TRAIN-SCORE-CROSSDEVICE).
 *
 * One entry per `drill_solves` row with `solved_at IS NOT NULL`, in
 * `position` order. The client aggregates these with `scorePuzzle` +
 * `aggregateSessionScore` from `@/lib/trainScore` — that module stays the
 * single source of truth for scoring (LOCKED, Option B); this response
 * deliberately carries NO precomputed score integer.
 */
export interface SolvedResult {
  correct_guess: boolean;
  move_quality: TrainMoveTier;
  /**
   * Phase 222 (D-17): the same three values `SolveResponse` already returned
   * for this exact position at the moment it was attempted — NOT a new
   * answer-key leak (mirrors this interface's own docstring rule above).
   * Entries here still carry no `position`, `game_id`, `ply` or best-move
   * field. `source` mirrors `SolveResponse.source` (required); `item_status`
   * and `due_date` mirror `SolveResponse.item_status`/`due_date` (nullable —
   * `null` for `red_herring`/`sharp_filler` sources).
   */
  source: 'sr_item' | 'red_herring' | 'sharp_filler';
  item_status: 'active' | 'mastered' | 'parked' | null;
  due_date: string | null;
}

/**
 * Response for POST /train/sessions — a composed or resumed session.
 *
 * `solved_results` (260728-tgc) is what makes "Scored today" correct on a
 * device that never saw the original solve responses — see `SolvedResult`'s
 * docstring. Empty for a freshly composed session and for the
 * no-eligible-material (`session_id === null`) case.
 *
 * `is_warmup` (Phase 206, D-06/D-07) is frozen at composition and derived
 * purely from material scarcity — true iff the session contains zero
 * surviving SR_ITEM puzzles at the moment it was (re-)composed. Never
 * derived from session ordinal, session count, or account age. The client
 * performs no arithmetic over it — a single equality read (T-191-24).
 */
export interface TrainSessionResponse {
  session_id: number | null;
  session_date: string;
  expires_on: string;
  puzzle_count: number;
  requested_count: number;
  solved_count: number;
  blob_pending_count: number;
  puzzles: TrainPuzzle[];
  solved_results: SolvedResult[];
  is_warmup: boolean;
}

/**
 * Body for POST /train/sessions/{session_id}/solve.
 *
 * P-02 (LOCKED) / SEED-119: the client asserts the three-way `move_quality`
 * tier (grading happens entirely client-side, SOLV-03) but NEVER
 * `correct_guess` — that is computed server-side from the live blob
 * classifier. The server derives the ladder verdict as
 * `move_quality != "wrong"`.
 */
export interface SolveRequest {
  position: number;
  guess: 'critical' | 'several';
  /** UCI move string: 4 chars normal ("e2e4"), 5 chars promotion ("e7e8q"). */
  played_move: string;
  move_quality: TrainMoveTier;
}

/**
 * Response for POST /train/sessions/{session_id}/solve.
 *
 * SEED-119: `correct_move` retains its exact prior meaning — the SR
 * ladder's pass/fail verdict, also what the reveal's check/cross mark
 * reads. `move_quality` is the new three-way scoring tier the client's
 * points formula consumes.
 *
 * Phase 206 (RESEARCH Pitfall 1): `source` mirrors `puzzle_type` — it lands
 * synchronously with this response, so the D-19 your-game predicates in
 * TrainReveal.tsx read `verdict.source`, never the separate, asynchronously
 * fetched `PuzzleRevealResponse.source` (which would open a post-solve
 * window where a real SR puzzle misrenders as suppressed).
 */
/**
 * One server-certified "also fine" alternative move (Phase 211, D-01) — the
 * wire mirror of `app.schemas.train.VettedMove`. Structurally identical to
 * `trainArrows.ts`'s `TrainFineMove`, on purpose: the reveal overlay consumes
 * these entries directly. D-01 amendment (2026-08-16): 'best' marks the deep
 * best move itself, served first on a soft puzzle alongside the certified
 * second-best, so the "Also fine" row always has a displayable entry after
 * the overlay filters out the client's own best/played arrows.
 */
export interface VettedMove {
  uci: string;
  quality: 'best' | 'good' | 'inaccuracy';
}

export interface SolveResponse {
  correct_guess: boolean;
  correct_move: boolean;
  move_quality: TrainMoveTier;
  puzzle_type: 'sharp' | 'soft' | 'herring';
  source: 'sr_item' | 'red_herring' | 'sharp_filler';
  item_status: 'active' | 'mastered' | 'parked' | null;
  streak: number | null;
  due_date: string | null;
  session_complete: boolean;
  /**
   * Phase 211 (D-01/D-03/D-07): the server's certified "also fine" set and
   * the graded-eval pair for a key-move override (non-null exactly when the
   * server overrode the client's tier). All three fields are OPTIONAL on
   * purpose: a `trainRevealCache` entry written by a pre-211 bundle restores
   * a `verdict` object with none of these keys at runtime even though the
   * compiler sees them — the same D-10 graceful-degradation pattern
   * `GradeResult.lines?` established. Exactly one nullish default per seam,
   * at the consumption site.
   */
  vetted_moves?: VettedMove[];
  graded_es_before?: number | null;
  graded_es_after?: number | null;
}

/**
 * Response for GET /train/sessions/{session_id}/puzzles/{position}/reveal.
 * Reachable ONLY after the attempt is recorded (409 otherwise — T-189-17).
 *
 * 190.1-03 (D-01/D-05): deliberately thin — NO `best_move`/`best_move_san`/
 * `pv` fields. The best move, the best line, and every eval shown in the
 * reveal panel are computed CLIENT-SIDE by the grading engine
 * (`useTrainGradingEngine.ts`), never derived or stored server-side (a
 * server-stored Stockfish eval and the client's own WASM search are not
 * guaranteed to agree bit-for-bit — project_eval_nondeterminism). Do not
 * re-add these fields as a "fallback" — see 190.1-03-PLAN.md's
 * assumption-delta decision.
 *
 * `played_in_game_move_uci` (190.1-01, D-05) is the UCI counterpart of
 * `played_in_game_san`, behind the identical gate — used to dispatch the
 * client's reveal-time engine search.
 *
 * `game_id` is `number | null` (Phase 192 Plan 02/05, D-05/D-09): a red
 * herring's source game can be nulled by the OWNER deleting it (a global
 * pool row survives independently, D-03) — never by the solving user. A
 * null `game_id` degrades the reveal's game footer to nothing (D-07) and
 * hides the Analyze deep-link (D-09) rather than disabling it.
 */
export interface PuzzleRevealResponse {
  game_id: number | null;
  ply: number;
  fen: string;
  played_in_game_san: string | null;
  played_in_game_move_uci: string | null;
  puzzle_type: 'sharp' | 'soft' | 'herring';
  source: 'sr_item' | 'red_herring' | 'sharp_filler';
  has_tactic_lines: boolean;
  /** Phase 206 (D-20): the sharp filler's motif label ("Fork", "Skewer",
   * ...), read straight from the committed data file. null for every
   * sr_item/red_herring reveal — no motif taxonomy exists for those today. */
  motif: string | null;
}

/** Response for GET /train/settings.
 *
 * `reminder_enabled`/`reminder_hour` (Phase 202, PERM-01..04) were exposed by
 * 201 D-18 but never round-tripped through the frontend until now. The
 * backend CHECK bound on `reminder_hour` is 0..23 inclusive
 * (`REMINDER_HOUR_MIN`/`REMINDER_HOUR_MAX` in `app/services/train_scheduler.py`).
 *
 * `reminder_intent_at` (Phase 203, OFFER-03/OFFER-05/D-02/D-15) is an ISO
 * instant string, or null if the user has never expressed install intent.
 */
export interface TrainSettingsResponse {
  timezone: string;
  weekday_mask: number;
  puzzles_per_session: number;
  reminder_enabled: boolean;
  reminder_hour: number;
  reminder_intent_at: string | null;
  /**
   * Phase 222 (D-11/D-12): server-stamped, read-only "seen" watermarks for
   * the three onboarding steppers — stamped by `POST /train/onboarding/
   * {step}` only on a COMPLETED stepper (an abandoned one replays). Never
   * writable via `TrainSettingsUpdate` (D-12) — a PUT body can never smuggle
   * a server-owned field, same rule as `reminder_last_sent_on`. `null` for
   * every existing row (D-13: no backfill) until the client stamps it once.
   */
  intro_seen_at: string | null;
  reveal_walkthrough_seen_at: string | null;
  sr_explained_at: string | null;
  /**
   * Phase 222 UAT round 5: whether ANY of the account's push subscriptions
   * came from a mobile browser (server-side User-Agent check). Account-wide,
   * unlike the per-device subscription probe — the desktop score screen uses
   * it to swap the phone QR handoff for a "reminders on your phone" line.
   */
  has_mobile_subscription: boolean;
}

/** Body for PUT /train/settings — a separate shape so a PUT can never smuggle
 * a server-owned field (mirrors the backend's `TrainSettingsUpdate`).
 *
 * This is a full-replace body with exactly one call site
 * (`useTrainSettings.ts`'s `mutationFn`) — every field here must be sent on
 * every save, or an existing save 422s the moment the backend requires it.
 * `reminder_intent_at` is required-but-nullable (D-02) — an ISO instant
 * string or null, never omitted. */
export interface TrainSettingsUpdate {
  timezone: string;
  weekday_mask: number;
  puzzles_per_session: number;
  reminder_enabled: boolean;
  reminder_hour: number;
  reminder_intent_at: string | null;
}

/**
 * The server-computed PROG-05/D-16 empty-state discriminant (mirrors the
 * backend `_pool_state`). `'no_material'` = never had any qualifying
 * material (cold start); `'exhausted'` = material existed but nothing is
 * waiting and nothing is still analyzing; `'available'` = every other case,
 * including a zero-`drill_items` user whose blunders are still being
 * analyzed ("catching up", not a cold start). The client performs no
 * arithmetic to pick between the two empty states — this field is the
 * single source of truth.
 */
export type TrainPoolState = 'no_material' | 'exhausted' | 'available';

/**
 * Response for GET /train/progress (PROG-01/PROG-04).
 *
 * Phase 193 replaced the Phase 191 weekly D-18 settled-streak snapshot with
 * a per-scheduled-day tick + a 0-7 depletable shield: `session_streak_count`
 * (was `settled_streak_weeks`) and `shield_level` (was `flame_state`, a
 * 3-state enum; now a plain number) mirror the persisted tick snapshot on
 * `train_settings`. There is no display overlay any more — the returned
 * values are always exactly what is persisted. `current_week_required` is
 * null when `weekday_mask === 0` ("train anytime" has no denominator to
 * show). `mastered_count`/`parked_count` are computed on the fly from
 * `drill_items` (D-05, unaffected by the tick snapshot).
 *
 * `waiting_count` is an upper-bound attention-signal estimate, never a
 * promise of exact session size — the start screen's own "N puzzles
 * waiting" line still comes from the real composed session. `next_due_date`
 * is the earliest date an ACTIVE item will next resurface, or null when
 * nothing will (the "All caught up!" empty state's date).
 *
 * `streak_reset_notice` (was `streak_lost_last_week`) is derived from the
 * RESULTING state (never from "did this call settle the reset"), so it
 * survives a page reload.
 *
 * `badge_visible` (Plan 02, D-09/D-10) is a DISPLAY HINT ONLY — it gates no
 * server-side authorization, and the number the nav badge shows still comes
 * from `waiting_count`. The client MUST NEVER attempt its own day-of-week or
 * timezone math here — it has no `weekday_mask` and no clean way to
 * reproduce the backend's `local_today` — this field is the single source
 * of truth for whether the badge should show.
 */
export interface TrainProgressResponse {
  session_streak_count: number;
  shield_level: number;
  current_week_completed: number;
  current_week_required: number | null;
  streak_reset_notice: boolean;
  mastered_count: number;
  parked_count: number;
  waiting_count: number;
  pool_state: TrainPoolState;
  next_due_date: string | null;
  badge_visible: boolean;
}

File: app/schemas/train.py
"""Pydantic v2 schemas for the Train API (Phase 189).

POOL-10 / P-01: `TrainPuzzle` is the pre-attempt payload and carries no
answer key — see its class docstring for the exact-equality contract this
schema exists to enforce. Phase 211: vetted-move material (`VettedMove`,
`SolveResponse.vetted_moves`, the graded-ES pair) lives on `SolveResponse`
— produced only once the attempt is already recorded — and never on
`TrainPuzzle`.
"""

from __future__ import annotations

from datetime import date, datetime
from typing import Literal
from zoneinfo import ZoneInfo, ZoneInfoNotFoundError

from pydantic import BaseModel, Field, field_validator

from app.services.train_scheduler import REMINDER_HOUR_MAX, REMINDER_HOUR_MIN


class TrainPuzzle(BaseModel):
    """One pre-attempt puzzle.

    POOL-10 / P-01 (LOCKED): the pre-attempt payload carries no answer key.
    `last_move_uci` (190-02, SOLV-02) describes the position's ARRIVAL — the
    half-move immediately before `ply`, i.e. the opponent's (or the user's
    own) prior move — so the solve screen can animate/highlight it. It does
    not reveal what to play next, so it does not reopen POOL-10. `best_move`,
    `pv`, `puzzle_type`, and `source` remain forbidden: adding any of them
    here re-opens the POOL-10 leak this schema exists to close — the
    client's exact-match/grading path runs entirely client-side against its
    own vendored Stockfish WASM output (see 189-01-PLAN.md P-01). Do not add
    fields here without re-reading that decision.

    `game_id` is `int | None` (Phase 192 Plan 02, D-01/D-05): a red herring's
    source-game link is nullable provenance — `None` here means either the
    herring's source game has been deleted (the puzzle is still fully
    servable off its `herring_pool` row, D-03) or the pool row was never
    linked to a game in the first place. Never used as an identity key
    client-side; a puzzle's identity within a session is `position`.
    """

    position: int
    game_id: int | None
    ply: int
    fen: str
    side_to_move: Literal["white", "black"]
    last_move_uci: str | None


class SolvedResult(BaseModel):
    """One recorded solve's outcome, part of `TrainSessionResponse.solved_results`.

    Quick task 260728-tgc (BUGFIX-TRAIN-SCORE-CROSSDEVICE): one entry per
    `drill_solves` row with `solved_at IS NOT NULL`, in `position` order. The
    client aggregates these with its own points formula
    (`frontend/src/lib/trainScore.ts`, `scorePuzzle` + `aggregateSessionScore`)
    — that file stays the single source of truth for scoring (LOCKED, Option
    B). This response deliberately carries NO precomputed score integer;
    porting the formula server-side was considered and rejected (see
    `app.models.drill_solve.DrillMoveQuality`'s docstring, which explicitly
    forbids using its enum values to compute a score directly).

    Not an answer-key leak: `correct_guess`, `move_quality`, and (Phase 222,
    TRAINBOT-04/D-17) `source`/`item_status`/`due_date` were ALL already
    returned by `SolveResponse` for each of these same positions at the
    moment they were attempted — this endpoint just re-serves outcomes the
    client already saw once, from the server instead of a device-local
    cache. The `PuzzleRevealResponse` 409 gate (which protects the actual
    answer key — best move, PV, puzzle type) is untouched and continues to
    protect UNSOLVED positions only. Entries here carry no `position`,
    `game_id`, `ply`, or best-move field, so they reveal nothing about
    puzzles still to be attempted in the session.

    `due_date` (D-17 caveat): this is the item's CURRENT value, not a frozen
    at-solve-time snapshot — identical within one session, and arguably more
    truthful for a "when does it come back" statement across a re-composed
    one (e.g. after a later solve in the same session re-advanced the ladder).
    `item_status`/`due_date` are `None` for `source in ("red_herring",
    "sharp_filler")`, which carry no SR bookkeeping (mirrors `SolveResponse`'s
    own nullability).
    """

    correct_guess: bool
    move_quality: Literal["good", "inaccuracy", "wrong"]
    source: Literal["sr_item", "red_herring", "sharp_filler"]
    item_status: Literal["active", "mastered", "parked"] | None
    due_date: date | None


class TrainSessionResponse(BaseModel):
    """Response for POST /train/sessions — a composed or resumed session.

    `session_id` is nullable: per the repository's explicit "write NO
    drill_sessions row when nothing qualifies" contract, a request that finds
    no eligible puzzle returns `session_id=None`, `puzzle_count=0`,
    `puzzles=[]` rather than a persisted empty session.

    `requested_count` is the settings value of N (`puzzles_per_session`);
    `blob_pending_count` is the number of the user's own qualifying blunders
    still waiting on opportunistic tier-4 analysis to populate their answer
    key. A caller seeing `puzzle_count < requested_count` MUST read
    `blob_pending_count` to tell "still analyzing" (non-zero) apart from
    "genuinely caught up" (zero) — removing either field re-hides the
    Pitfall 4 signal this schema exists to surface.

    `solved_results` (260728-tgc) is one `SolvedResult` per recorded solve in
    `position` order — see that schema's docstring. Empty for a freshly
    composed session and for the no-eligible-material (`session_id is None`)
    case. This is what makes "Scored today" correct on a device that never
    saw the original solve responses (the reproduced prod bug: a
    localStorage-only tally read "0 of 18" on a second device).

    `is_warmup` (Phase 206, D-06/D-07) is frozen at composition and derived
    purely from material scarcity — `True` iff the session contains zero
    surviving SR_ITEM puzzles at the moment it was (re-)composed. It is
    never derived from session ordinal, session count, or account age, and
    the client performs no arithmetic to reproduce it: it is a single
    server-computed boolean to branch on (T-191-24).
    """

    session_id: int | None
    session_date: date
    expires_on: date
    puzzle_count: int
    requested_count: int
    solved_count: int
    blob_pending_count: int
    puzzles: list[TrainPuzzle]
    solved_results: list[SolvedResult]
    is_warmup: bool


class SolveRequest(BaseModel):
    """Body for POST /train/sessions/{session_id}/solve.

    P-02 (LOCKED) / SEED-119: the client asserts a three-way `move_quality`
    tier (the backend never grades the move — grading is still entirely
    client-side, see the module/plan docstrings) but NEVER `correct_guess`
    or `puzzle_type` — those are computed server-side from the live
    `game_flaws` blob so the sharp/soft ground truth is never handed to the
    client before the attempt (T-189-18/T-189-11). The server derives the
    spaced-repetition ladder's pass/fail boolean from `move_quality`
    (`!= "wrong"`) — see `app.repositories.train_repository.record_solve`.

    Phase 211 (D-03/D-07) narrows the sentence above: as of this phase the
    backend DOES grade the move for the one case where it owns the truth —
    a `played_move` matching the server-certified key (the soft blob's `su`,
    or a qualifying herring ladder entry). The client's assertion is still
    required in every request and still authoritative for every OFF-key move
    (D-04's accepted residual); for a key move the server recomputes the
    tier from its own stored evals and discards the client's value. This is
    the same shape as the pre-existing `_compute_correct_guess` override —
    the client can never assert a verdict it does not own. The request
    schema itself is unchanged (P-01 intact: the client cannot know the key
    before attempting).
    """

    position: int
    guess: Literal["critical", "several"]
    # UCI move string: 4 chars normal (e.g. "e2e4"), 5 chars promotion (e.g. "e7e8q").
    played_move: str = Field(min_length=4, max_length=5)
    move_quality: Literal["good", "inaccuracy", "wrong"]


class VettedMove(BaseModel):
    """One server-certified "also fine" alternative move (Phase 211, D-01).

    The wire twin of `app.services.train_pool.VettedMove` (same name,
    different module — mapped field-by-field at the router). Deliberately
    thin: only the UCI and its pre-classified quality tier cross the wire;
    the underlying expected scores stay server-side (the graded-ES pair on
    `SolveResponse` covers the one move that needs numbers client-side).

    POST-ATTEMPT-ONLY material: this model appears exclusively on
    `SolveResponse`, which is produced solely by `record_solve` after the
    attempt row is resolved. Never add it to `TrainPuzzle` (P-01) or
    `SolveRequest`.

    D-01 amendment (2026-08-16): quality "best" marks the deep best move
    itself, served first on a soft puzzle alongside the certified
    second-best (`game_positions.best_move` supplies the UCI; su-only when
    unavailable) so the "several fine moves" copy is always backed by a
    displayable alternative after the client filters its own best/played
    arrows out of the row.
    """

    uci: str
    quality: Literal["best", "good", "inaccuracy"]


class SolveResponse(BaseModel):
    """Response for POST /train/sessions/{session_id}/solve.

    `item_status`/`streak`/`due_date` are None for a red-herring puzzle,
    which carries no SR bookkeeping (POOL-08). `correct_guess` is always the
    server-computed verdict, never an echo of the client's own guess.

    SEED-119: `correct_move` retains its exact prior meaning — the
    spaced-repetition ladder's pass/fail verdict, which is also what the
    reveal's check/cross mark reads. `move_quality` is the new three-way
    scoring tier the client's points formula consumes; it is NOT a synonym
    for `correct_move` (an "inaccuracy" tier still means `correct_move=True`).

    Phase 206 (RESEARCH Pitfall 1): `source` mirrors `puzzle_type`'s existing
    shape and lands synchronously with this response — the client's D-19
    your-game predicates read `verdict.source`, never the separate,
    asynchronously-fetched `PuzzleRevealResponse.source`, to avoid a
    post-solve window where a real SR puzzle misrenders as suppressed.

    Phase 211 (D-01/D-03/D-07, amended 2026-08-16): `vetted_moves` is the
    server's certified "also fine" set for this puzzle — the deep best plus
    the blob's `su` for a soft puzzle (best-first; su-only when
    `game_positions.best_move` is unavailable), the good-band ladder entries
    for a herring, always empty for sharp/sharp-filler — and is what the
    reveal's legend row + green arrows render (never the client engine's own
    alternatives). `graded_es_before`
    / `graded_es_after` are non-null EXACTLY when this call claimed the row
    AND the played move matched a vetted entry (the server override): they
    carry the same mover-POV expected scores the overridden `move_quality`
    was computed from, so the client can re-classify the board badge from
    the SAME numbers the score came from — display and recorded verdict can
    never diverge for a key move.
    """

    correct_guess: bool
    correct_move: bool
    move_quality: Literal["good", "inaccuracy", "wrong"]
    puzzle_type: Literal["sharp", "soft", "herring"]
    source: Literal["sr_item", "red_herring", "sharp_filler"]
    item_status: Literal["active", "mastered", "parked"] | None
    streak: int | None
    due_date: date | None
    session_complete: bool
    vetted_moves: list[VettedMove]
    graded_es_before: float | None
    graded_es_after: float | None


class PuzzleRevealResponse(BaseModel):
    """Response for GET /train/sessions/{session_id}/puzzles/{position}/reveal.

    Reachable ONLY after the attempt is recorded (409 otherwise — T-189-17):
    the puzzle type and the in-game move are unreachable before `solved_at`
    is set.

    190.1-03 (D-01/D-05): this response is DELIBERATELY thin. The answer key
    it carries is the puzzle type, the in-game move (SAN + UCI), and a
    tactic-lines pointer — no `best_move`, `best_move_san`, or `pv` field.
    The best move, the best line, and every eval shown in the reveal panel
    are computed CLIENT-SIDE by the grading engine (`useTrainGradingEngine.ts`),
    never derived or stored here — a server-stored Stockfish eval and the
    client's own WASM search are not guaranteed to agree bit-for-bit
    (project_eval_nondeterminism), so this endpoint must never be a second,
    contradicting source of truth for a number the reveal panel displays.

    `played_in_game_move_uci` (190.1-01, D-05) is the UCI counterpart of
    `played_in_game_san`, behind the identical 409 gate — the client uses it
    to dispatch its own reveal-time engine search (T-190.1-01/T-190.1-02).

    `has_tactic_lines` is a POINTER, not a payload: when True, the client
    calls the existing `GET /api/library/flaws/{game_id}/{ply}/tactic-lines`
    endpoint for the steppable PV line. Train adds no second PV-fetching
    surface — see 189-05-PLAN.md's key_links.

    `game_id` is `int | None` (Phase 192 Plan 02, D-01/D-05): `None` means the
    puzzle's source game has since been deleted. The client hides the Analyze
    deep-link in that case (D-09) rather than disabling it — nothing else on
    the reveal panel references the game either way.
    """

    game_id: int | None
    ply: int
    fen: str
    played_in_game_san: str | None
    played_in_game_move_uci: str | None
    puzzle_type: Literal["sharp", "soft", "herring"]
    source: Literal["sr_item", "red_herring", "sharp_filler"]
    has_tactic_lines: bool
    # Phase 206 (D-20): the sharp filler's motif label. None for every
    # SR_ITEM/RED_HERRING row.
    motif: str | None


#: Phase 222 (TRAINBOT-05, D-11/D-12): the path parameter for
#: POST /train/onboarding/{step}. FastAPI 422s any value outside this set
#: before the handler body runs — no hand-rolled validation, and the
#: repository maps each member through a fixed column lookup (never a
#: string interpolated into SQL, never `getattr` on a request-supplied name).
OnboardingStep = Literal["intro", "reveal_walkthrough", "sr_explained"]


class TrainSettingsResponse(BaseModel):
    """Response for GET/PUT /train/settings.

    `reminder_enabled`/`reminder_hour` (Phase 201, REMIND-01/D-18) are the
    user-owned reminder configuration, round-tripped through this same
    settings surface so it is fully testable before Phase 202 builds any UI.
    `reminder_last_sent_on` is deliberately absent — it is the reminder job's
    own watermark (D-06), never readable or writable by a client.
    `reminder_intent_at` (Phase 203, OFFER-03/OFFER-05/D-02/D-15) is the
    OPPOSITE of `reminder_last_sent_on`: it IS client-writable (stamped when
    the user taps the iOS install affordance), so it appears here.

    `intro_seen_at`/`reveal_walkthrough_seen_at`/`sr_explained_at` (Phase 222,
    TRAINBOT-05, D-11/D-12/D-13) are the three bot-onboarding "explanation
    seen" watermarks. Like `reminder_last_sent_on`, each is a server-owned
    fact — but unlike it, all three ARE readable here (so the client can
    branch on them without a second fetch) while remaining absent from
    `TrainSettingsUpdate`: they are stamped ONLY by the dedicated
    `POST /train/onboarding/{step}` endpoint on stepper completion, never by
    a settings PUT. No backfill (D-13): every row that predates these
    columns reads back NULL on all three, meaning "never seen."

    `has_mobile_subscription` (Phase 222 UAT round 5) is a derived, read-only
    fact: whether ANY of the account's push subscriptions came from a mobile
    browser (`push_repository.MOBILE_USER_AGENT_PATTERN`). The desktop score
    screen uses it to replace the phone QR handoff with a "reminders on your
    phone" line — a per-device subscription probe can never answer that.
    """

    timezone: str
    weekday_mask: int
    puzzles_per_session: int
    reminder_enabled: bool
    reminder_hour: int
    reminder_intent_at: datetime | None
    intro_seen_at: datetime | None
    reveal_walkthrough_seen_at: datetime | None
    sr_explained_at: datetime | None
    has_mobile_subscription: bool


class TrainSettingsUpdate(BaseModel):
    """Body for PUT /train/settings.

    A separate schema from `TrainSettingsResponse` (not one schema reused for
    both directions) so a PUT body can never smuggle a server-owned field.
    `weekday_mask`/`puzzles_per_session`/`reminder_hour` bounds mirror the
    `train_settings` table's CHECK constraints exactly. `reminder_last_sent_on`
    is deliberately absent for the same reason it is absent from
    `TrainSettingsResponse` — see that class's docstring.

    `reminder_intent_at` has NO default (required-but-nullable), a deliberate
    choice per D-02: this schema is a full-replace PUT body, so a defaulted
    field would let a payload that omits the key silently clear a
    previously-recorded install intent. Omitting the key must 422 loudly
    instead. No bound validator: a timestamp has no injection surface
    (RESEARCH.md Security Domain V5).
    """

    timezone: str
    weekday_mask: int = Field(ge=0, le=127)
    puzzles_per_session: int = Field(ge=1, le=50)
    reminder_enabled: bool
    reminder_hour: int = Field(ge=REMINDER_HOUR_MIN, le=REMINDER_HOUR_MAX)
    reminder_intent_at: datetime | None

    @field_validator("timezone")
    @classmethod
    def _validate_timezone(cls, value: str) -> str:
        """D-06: reject an unresolvable IANA timezone with 422, never persist it.

        A stored bad zone would silently shift every future due-date and
        session-window computation (`local_today` falls back to UTC for a
        legacy bad value already on a row, but nothing new may be written
        that way).
        """
        try:
            ZoneInfo(value)
        except (ZoneInfoNotFoundError, ValueError) as exc:
            raise ValueError(f"Unrecognized IANA timezone: {value!r}") from exc
        return value


class TrainProgressResponse(BaseModel):
    """Response for GET /train/progress (PROG-01/PROG-04).

    Phase 193 (SEED-121) replaced Phase 191's weekly D-18 settled-streak
    snapshot with a per-scheduled-day tick + a 0-7 depletable shield:
    `session_streak_count` (was `settled_streak_weeks`; the wire field spells
    out the D-02 unit — it counts completed scheduled-day SESSIONS, not
    settled weeks) and `shield_level` (was `flame_state`, a 3-state enum;
    now a plain int) come from the persisted tick snapshot on
    `train_settings`, lazily advanced by this same request
    (`app.repositories.train_repository.settle_streak_snapshot`). There is
    no display overlay any more — the returned values are always exactly
    what is persisted. `current_week_required` is None when
    `weekday_mask == 0` ("train anytime" has no denominator to show);
    otherwise it is the popcount of the scheduled-day mask (no special-
    casing — nothing gates on this value any more).
    `mastered_count`/`parked_count` are computed on the fly from
    `drill_items` (D-05, unaffected by the tick snapshot — only the
    streak/shield portion is persisted).

    `waiting_count`/`pool_state`/`next_due_date` are the server-side signals
    the nav badge and the two PROG-05 empty states need: `waiting_count` is
    an upper-bound estimate of puzzles waiting right now (never a promise of
    exact session size — see
    `app.repositories.train_repository.get_waiting_puzzle_count`).
    `pool_state` is the single discriminant the client branches on for the
    empty states: `"no_material"` means the user has never had any
    qualifying material (cold start); `"exhausted"` means material existed
    but nothing is waiting and nothing is still analyzing; `"available"`
    covers every other case, including a zero-`drill_items` user whose own
    blunders are still being analyzed (that is "catching up", not a cold
    start). `next_due_date` is the earliest date an ACTIVE item will next
    resurface, or null when nothing will (the "All caught up!" empty state's
    date).

    `streak_reset_notice` (was `streak_lost_last_week`) is derived from the
    RESULTING state (never from "did this call settle the reset"), so it
    survives a page reload and self-clears once the user trains again.

    `badge_visible` (Plan 02, D-09/D-10) is a DISPLAY HINT ONLY — it gates no
    server-side authorization, and the number the nav badge shows still
    comes from `waiting_count`. True when `waiting_count > 0` AND (today is
    a scheduled day per the user's `weekday_mask` OR an already-open
    unexpired session still has unsolved puzzles left to rescue). The client
    performs no day-of-week or timezone math of its own — it has no
    `weekday_mask` and no clean way to reproduce `local_today`, so this
    field is the single source of truth for whether the badge should show.
    """

    session_streak_count: int
    shield_level: int
    current_week_completed: int
    current_week_required: int | None
    streak_reset_notice: bool
    mastered_count: int
    parked_count: int
    waiting_count: int
    pool_state: Literal["no_material", "exhausted", "available"]
    next_due_date: date | None
    badge_visible: bool


__all__ = [
    "PuzzleRevealResponse",
    "SolveRequest",
    "SolveResponse",
    "SolvedResult",
    "TrainProgressResponse",
    "TrainPuzzle",
    "TrainSessionResponse",
    "TrainSettingsResponse",
    "TrainSettingsUpdate",
    "VettedMove",
]
